doodles/Doodles.sol

```diff
@@ -0,0 +1,107 @@
+        for (uint256 i = 0; i < addresses.length; i++) {
+            _allowList[addresses[i]] = numAllowedToMint;
+        }
+    }
+
+    // ## this is a read only function that anyone can execute. Since it is a view it requires no gas.  Returns the amount of items you can mint if you are white listed
+    function numAvailableToMint(address addr) external view returns (uint8) {
+        return _allowList[addr];
+    }
+
+    // ## allows folks to mint as long as they are on the allow list and meet someother basic critera
+    function mintAllowList(uint8 numberOfTokens) external payable {
+        uint256 ts = totalSupply();
+        require(isAllowListActive, "Allow list is not active");
+        require(numberOfTokens <= _allowList[msg.sender], "Exceeded max available to purchase");
+        require(ts + numberOfTokens <= MAX_SUPPLY, "Purchase would exceed max tokens");
+        require(PRICE_PER_TOKEN * numberOfTokens <= msg.value, "Ether value sent is not correct");
+
+        // ## decrement the amount of mints the sender can do
+        _allowList[msg.sender] -= numberOfTokens;
+
+        // ## loop through the number of tokens
+        for (uint256 i = 0; i < numberOfTokens; i++) {
+            _safeMint(msg.sender, ts + i);
+        }
+    }
+
+    function _beforeTokenTransfer(address from, address to, uint256 tokenId) internal override(ERC721, ERC721Enumerable) {
+        super._beforeTokenTransfer(from, to, tokenId);
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC721, ERC721Enumerable) returns (bool) {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function setBaseURI(string memory baseURI_) external onlyOwner() {
+        _baseURIextended = baseURI_;
+    }
+
+    function _baseURI() internal view virtual override returns (string memory) {
+        return _baseURIextended;
+    }
+
+    function setProvenance(string memory provenance) public onlyOwner {
+        PROVENANCE = provenance;
+    }
+
+    function reserve(uint256 n) public onlyOwner {
+      uint supply = totalSupply();
+      uint i;
+      for (i = 0; i < n; i++) {
+          _safeMint(msg.sender, supply + i);
+      }
+    }
+
+    // ## setting this to true will allow every joe schmoe to mint
+    function setSaleState(bool newState) public onlyOwner {
+        saleIsActive = newState;
+    }
+
+    function mint(uint numberOfTokens) public payable {
+        uint256 ts = totalSupply();
+        require(saleIsActive, "Sale must be active to mint tokens");
+        require(numberOfTokens <= MAX_PUBLIC_MINT, "Exceeded max token purchase");
+        require(ts + numberOfTokens <= MAX_SUPPLY, "Purchase would exceed max tokens");
+        require(PRICE_PER_TOKEN * numberOfTokens <= msg.value, "Ether value sent is not correct");
+
+        for (uint256 i = 0; i < numberOfTokens; i++) {
+            _safeMint(msg.sender, ts + i);
```

Comment:
I think I have a good understanding of how the minting for this contract works. What I can't seem to piece together is how is the metadata attached to the tokenId?

Reply:
tokenURI takes a token id as input and spits back a URI to the metadata. 

this is built by convention. 

the baseURI is typically:

ipfs://hwdjh6jsjfk86hsjkfkshxj/

and then to reference a specific token’s metadata would be something like:

ipfs://hwdjh6jsjfk86hsjkfkshxj/25

where 25 is token id 25’s metadata json